crates/amaru/src/stages/pull/mod.rs

```diff
@@ -12,37 +12,56 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use crate::send;
+use std::sync::Arc;
+
+use crate::{
+    send,
+    stages::{metrics::MetricsEvent, pull::metrics::PullMetrics},
+};
 use amaru_consensus::consensus::ChainSyncEvent;
 use amaru_kernel::{Point, peer::Peer};
 use amaru_network::{chain_sync_client::ChainSyncClient, point::from_network_point};
-use gasket::framework::*;
+use gasket::framework::{Stage as StageTrait, *};
 use pallas_network::miniprotocols::chainsync::{Client, HeaderContent, NextResponse, Tip};
 use tracing::{Level, Span, debug, instrument};
 
+pub mod metrics;
+
 pub type DownstreamPort = gasket::messaging::OutputPort<ChainSyncEvent>;
+pub type MetricsDownstreamPort = gasket::messaging::OutputPort<MetricsEvent>;
 
 pub enum WorkUnit {
     Pull,
     Await,
     Intersect,
 }
 
-#[derive(Stage)]
+#[derive(StageTrait)]
 #[stage(name = "stage.chain_sync_client", unit = "WorkUnit", worker = "Worker")]
 pub struct Stage {
     pub peer: Peer,
     pub client: ChainSyncClient,
     pub downstream: DownstreamPort,
+    pub metrics_downstream: Option<MetricsDownstreamPort>,
 }
 
 impl Stage {
-    pub fn new(peer: Peer, chain_sync: Client<HeaderContent>, intersection: Vec<Point>) -> Self {
+    pub fn new(
+        peer: Peer,
+        chain_sync: Client<HeaderContent>,
+        intersection: Vec<Point>,
+        send_metrics: bool,
+    ) -> Self {
         let client = ChainSyncClient::new(peer.clone(), chain_sync, intersection);
         Self {
             peer,
             client,
             downstream: Default::default(),
+            metrics_downstream: if send_metrics {
+                Some(MetricsDownstreamPort::default())
+            } else {
+                None
+            },
```

Comment:
This feels like the dependency injection is backward. I would suggest to take a function instead of a boolean; that is invoked no matter what; but can be instantiated to a no-op by the caller, or to something different by the caller if needed.

Reply:
In general, let's take the time to make things right the first time. There is no rush here and I don't quite see why this should be considered PoC.
Those things usually are not changed once they are merged.

Reply:
Yes, of course. That is what we're doing here. I am just not doing micro optimizations that could have little to no impact.